modules/message-lane/src/lib.rs

```diff
@@ -90,19 +159,237 @@ decl_module! {
+					target: "runtime",
+					"Message to lane {:?} is rejected by target chain: {:?}",
+					lane_id,
+					err,
+				);
+
+				Error::<T, I>::MessageRejectedByChainVerifier
+			})?;
+
+			// now let's enforce any additional lane rules
+			T::LaneMessageVerifier::verify_message(
+				&submitter,
+				&delivery_and_dispatch_fee,
+				&lane_id,
+				&payload,
+			).map_err(|err| {
+				frame_support::debug::trace!(
+					target: "runtime",
+					"Message to lane {:?} is rejected by lane verifier: {:?}",
+					lane_id,
+					err,
+				);
+
+				Error::<T, I>::MessageRejectedByLaneVerifier
+			})?;
+
+			// let's withdraw delivery and dispatch fee from submitter
+			T::MessageDeliveryAndDispatchPayment::pay_delivery_and_dispatch_fee(
+				&submitter,
+				&delivery_and_dispatch_fee,
+			).map_err(|err| {
+				frame_support::debug::trace!(
+					target: "runtime",
+					"Message to lane {:?} is rejected because submitter {:?} is unable to pay fee {:?}: {:?}",
+					lane_id,
+					submitter,
+					delivery_and_dispatch_fee,
+					err,
+				);
+
+				Error::<T, I>::FailedToWithdrawMessageFee
+			})?;
+
+			// finally, save message in outbound storage and emit event
 			let mut lane = outbound_lane::<T, I>(lane_id);
-			let nonce = lane.send_message(payload);
+			let nonce = lane.send_message(MessageData {
+				payload,
+				fee: delivery_and_dispatch_fee,
+			});
 			lane.prune_messages(T::MaxMessagesToPruneAtOnce::get());
 
+			frame_support::debug::trace!(
+				target: "runtime",
+				"Accepted message {} to lane {:?}",
+				nonce,
+				lane_id,
+			);
+
 			Self::deposit_event(Event::MessageAccepted(lane_id, nonce));
+
+			Ok(())
+		}
+
+		/// Receive messages proof from bridged chain.
+		#[weight = DELIVERY_BASE_WEIGHT + max_dispatch_weight]
+		pub fn receive_messages_proof(origin, proof: MessagesProofOf<T, I>, max_dispatch_weight: Weight) -> DispatchResult {
+			let relayer = ensure_signed(origin)?;
+
+			// verify messages proof && convert proof into messages
+			let messages = T::SourceHeaderChain::verify_messages_proof(proof).map_err(|err| {
+				frame_support::debug::trace!(
+					target: "runtime",
+					"Rejecting invalid messages proof: {:?}",
+					err,
+				);
+
+				Error::<T, I>::InvalidMessagesProof
+			})?;
+
+			// estimate cumulative weight of messages dispatch
+			let mut dispatcher = T::MessageDispatch::with_allowed_weight(max_dispatch_weight);
+			let messages_dispatch_weight: Weight = messages.iter().map(|m| dispatcher.dispatch_weight(m)).sum();
+
+			// the submitter pays for dispatch in advance, so the whole `messages_dispatch_weight` should
+			// be covered. But it has already paid regular fee for at least `max_dispatch_weight`, so we
+			// only need to withdraw `messages_dispatch_weight - max_dispatch_weight`.
+			let messages_dispatch_extra_weight = messages_dispatch_weight.saturating_sub(max_dispatch_weight);
+			T::MessageDispatchPayment::pay_dispatch_fee(&relayer, messages_dispatch_extra_weight).map_err(|err| {
```

Comment:
IMHO this should use `Currency` trait directly, we shouldn't be abstracting the fees more than necessary.

Reply:
I imagine we should also charge the relayers who queue messages a tiny bit more, to prevent them from just sending lowest possible `max_dispatch_weight` and abusing  dispatch during `on_initialize`.

Reply:
> IMHO this should use Currency trait directly, we shouldn't be abstracting the fees more than necessary.

Responded in other comment. Here, imo, relying on `Currency` is even worse, because fee != currency. We may e.g. transfer fees to treasure, burn them, tip validators for processing queued messages, or even combine these actions.

> I imagine we should also charge the relayers who queue messages a tiny bit more, to prevent them from just sending lowest possible max_dispatch_weight and abusing dispatch during on_initialize.

In this case relayers would never deliver messages if there's anything queued in the lane. So this would be semantically equal to having no queued messages. As I said above, we may remove supporting messages queued for now => this should solve the issue. I'm just waiting for your opinion, guys.